In [51]:
import pandas as pd  #Importing necessary packages

In [53]:
#importing .csv file to work on
survey_responses = pd.read_csv('/Users/anthonydaniels/Desktop/Udacity Data Science Study/Udacity Project 1/Udacity Project 1 csv dataset 1.csv')

In [55]:
survey_responses.head()  #look at first 5 lines of code



ResponseId                      MainBranch                 Age  \
0           1  I am a developer by profession  Under 18 years old   
1           2  I am a developer by profession     35-44 years old   
2           3  I am a developer by profession     45-54 years old   
3           4           I am learning to code     18-24 years old   
4           5  I am a developer by profession     18-24 years old   

            Employment RemoteWork   Check  \
0  Employed, full-time     Remote  Apples   
1  Employed, full-time     Remote  Apples   
2  Employed, full-time     Remote  Apples   
3   Student, full-time        NaN  Apples   
4   Student, full-time        NaN  Apples   

                                    CodingActivities  \
0                                              Hobby   
1  Hobby;Contribute to open-source projects;Other...   
2  Hobby;Contribute to open-source projects;Other...   
3                                                NaN   
4                                                NaN   

                                             EdLevel  \
0                          Primary/elementary school   
1       Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
2    Master’s degree (M.A., M.S., M.Eng., MBA, etc.)   
3  Some college/university study without earning ...   
4  Secondary school (e.g. American high school, G...   

                                           LearnCode  \
0                             Books / Physical media   
1  Books / Physical media;Colleague;On the job tr...   
2  Books / Physical media;Colleague;On the job tr...   
3  Other online resources (e.g., videos, blogs, f...   
4  Other online resources (e.g., videos, blogs, f...   

                                     LearnCodeOnline  ... JobSatPoints_6  \
0                                                NaN  ...            NaN   
1  Technical documentation;Blogs;Books;Written Tu...  ...            0.0   
2  Technical documentation;Blogs;Books;Written Tu...  ...            NaN   
3  Stack Overflow;How-to videos;Interactive tutorial  ...            NaN   
4  Technical documentation;Blogs;Written Tutorial...  ...            NaN   

  JobSatPoints_7 JobSatPoints_8 JobSatPoints_9 JobSatPoints_10  \
0            NaN            NaN            NaN             NaN   
1            0.0            0.0            0.0             0.0   
2            NaN            NaN            NaN             NaN   
3            NaN            NaN            NaN             NaN   
4            NaN            NaN            NaN             NaN   

  JobSatPoints_11           SurveyLength SurveyEase ConvertedCompYearly JobSat  
0             NaN                    NaN        NaN                 NaN    NaN  
1             0.0                    NaN        NaN                 NaN    NaN  
2             NaN  Appropriate in length       Easy                 NaN    NaN  
3             NaN               Too long       Easy                 NaN    NaN  
4             NaN              Too short       Easy                 NaN    NaN  

[5 rows x 114 columns]

In [57]:
survey_responses.describe() #descriptive statistics of file

ResponseId      CompTotal       WorkExp  JobSatPoints_1  \
count  65437.000000   3.374000e+04  29658.000000    29324.000000   
mean   32719.000000  2.963841e+145     11.466957       18.581094   
std    18890.179119  5.444117e+147      9.168709       25.966221   
min        1.000000   0.000000e+00      0.000000        0.000000   
25%    16360.000000   6.000000e+04      4.000000        0.000000   
50%    32719.000000   1.100000e+05      9.000000       10.000000   
75%    49078.000000   2.500000e+05     16.000000       22.000000   
max    65437.000000  1.000000e+150     50.000000      100.000000   

       JobSatPoints_4  JobSatPoints_5  JobSatPoints_6  JobSatPoints_7  \
count    29393.000000    29411.000000    29450.000000     29448.00000   
mean         7.522140       10.060857       24.343232        22.96522   
std         18.422661       21.833836       27.089360        27.01774   
min          0.000000        0.000000        0.000000         0.00000   
25%          0.000000        0.000000        0.000000         0.00000   
50%          0.000000        0.000000       20.000000        15.00000   
75%          5.000000       10.000000       30.000000        30.00000   
max        100.000000      100.000000      100.000000       100.00000   

       JobSatPoints_8  JobSatPoints_9  JobSatPoints_10  JobSatPoints_11  \
count    29456.000000    29456.000000     29450.000000     29445.000000   
mean        20.278165       16.169432        10.955713         9.953948   
std         26.108110       24.845032        22.906263        21.775652   
min          0.000000        0.000000         0.000000         0.000000   
25%          0.000000        0.000000         0.000000         0.000000   
50%         10.000000        5.000000         0.000000         0.000000   
75%         25.000000       20.000000        10.000000        10.000000   
max        100.000000      100.000000       100.000000       100.000000   

       ConvertedCompYearly        JobSat  
count         2.343500e+04  29126.000000  
mean          8.615529e+04      6.935041  
std           1.867570e+05      2.088259  
min           1.000000e+00      0.000000  
25%           3.271200e+04      6.000000  
50%           6.500000e+04      7.000000  
75%           1.079715e+05      8.000000  
max           1.625660e+07     10.000000

In [59]:
survey_responses.isna().sum() #check for nulls and count them

ResponseId                 0
MainBranch                 0
Age                        0
Employment                 0
RemoteWork             10631
                       ...  
JobSatPoints_11        35992
SurveyLength            9255
SurveyEase              9199
ConvertedCompYearly    42002
JobSat                 36311
Length: 114, dtype: int64

In [61]:
survey_responses.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65437 entries, 0 to 65436
Columns: 114 entries, ResponseId to JobSat
dtypes: float64(13), int64(1), object(100)
memory usage: 56.9+ MB


We have 114 columns/features in our dataset.  We need to start the cleaning process.
Our project is going to attempt create a model that predicts those who will create a career which will be defined as those with over 10 years work experience.

In [29]:
# eliminate columns that we don't want to be taken into account for our model
# we have a column 'check' that appears to have responses all 'apples'.  check to see if there are any other responses

# Check for non-'apples' responses in the 'check' column
non_apples = survey_responses[survey_responses['Check'] != 'Apples']

# Display results
if len(non_apples) > 0:
    print(f"Found {len(non_apples)} rows with non-'apples' responses:")
    print(non_apples[['Check']])  # Show only the 'check' column
else:
    print("All responses in 'Check' column are 'Apples'.")

All responses in 'Check' column are 'Apples'.


Based on the fact that all of the responses are 'Apples' under the 'Check' column, we can assume that it is a nonimpactful column since it applies to everybody in the survey

In [63]:
#code to remove undesired columns from file
survey_responses.drop(columns=['Check', 'OrgSize', 'PurchaseInfluence', 'BuyNewTool', 'BuildvsBuy', 'TechEndorse', 'Currency','CompTotal', 'LanguageAdmired', 'DatabaseAdmired', 'PlatformAdmired', 'WebframeAdmired', 'EmbeddedAdmired', 'MiscTechWantToWorkWith', 'MiscTechAdmired', 'ToolsTechAdmired', 'NEWCollabToolsWantToWorkWith', 'NEWCollabToolsAdmired', 'OfficeStackAsyncWantToWorkWith', 'OfficeStackAsyncAdmired', 'OfficeStackSyncWantToWorkWith', 'OfficeStackSyncAdmired', 'AISearchDevWantToWorkWith', 'AISearchDevAdmired', 'SurveyLength', 'SurveyEase', 'ConvertedCompYearly'], inplace=True)

In [37]:
survey_responses.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65437 entries, 0 to 65436
Data columns (total 87 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   ResponseId                      65437 non-null  int64  
 1   MainBranch                      65437 non-null  object 
 2   Age                             65437 non-null  object 
 3   Employment                      65437 non-null  object 
 4   RemoteWork                      54806 non-null  object 
 5   CodingActivities                54466 non-null  object 
 6   EdLevel                         60784 non-null  object 
 7   LearnCode                       60488 non-null  object 
 8   LearnCodeOnline                 49237 non-null  object 
 9   TechDoc                         40897 non-null  object 
 10  YearsCode                       59869 non-null  object 
 11  YearsCodePro                    51610 non-null  object 
 12  DevType                         

In [69]:
WorkExp_NA = survey_responses['WorkExp'].isna().sum()  #determine if there are NA to be dealth with in WorkExp column
print (WorkExp_NA)

35779


In [73]:
survey_responses['Career'] = survey_responses['WorkExp'].fillna(0) >=10   #creates new column named Career and assigns False to everything w/NA or Less than 10 years of work experience

In [81]:
print(survey_responses[['WorkExp','Career']].head()) ##checks WorkExp vs new Career column

   WorkExp  Career
0      NaN   False
1     17.0    True
2      NaN   False
3      NaN   False
4      NaN   False
